data/src/test/java/org/apache/iceberg/io/TestGenericSortedPosDeleteWriter.java

```diff
@@ -0,0 +1,315 @@
+import org.apache.iceberg.data.parquet.GenericParquetReaders;
+import org.apache.iceberg.encryption.EncryptedOutputFile;
+import org.apache.iceberg.parquet.Parquet;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableList;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.util.StructLikeSet;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+
+@RunWith(Parameterized.class)
+public class TestGenericSortedPosDeleteWriter extends TableTestBase {
+  private static final int FORMAT_V2 = 2;
+
+  private final FileFormat format;
+
+  private OutputFileFactory fileFactory;
+  private Record gRecord;
+
+  @Parameterized.Parameters(name = "FileFormat={0}")
+  public static Object[] parameters() {
+    return new Object[][] {
+        new Object[] {"avro"},
+        new Object[] {"parquet"},
+    };
+  }
+
+  public TestGenericSortedPosDeleteWriter(String fileFormat) {
+    super(FORMAT_V2);
+    this.format = FileFormat.valueOf(fileFormat.toUpperCase(Locale.ENGLISH));
+  }
+
+  @Before
+  public void setupTable() throws IOException {
+    this.tableDir = temp.newFolder();
+    Assert.assertTrue(tableDir.delete());
+
+    this.metadataDir = new File(tableDir, "metadata");
+    this.table = create(SCHEMA, PartitionSpec.unpartitioned());
+    this.gRecord = GenericRecord.create(SCHEMA);
+
+    this.fileFactory = new OutputFileFactory(table.spec(), format, table.locationProvider(), table.io(),
+        table.encryption(), 1, 1);
+
+    table.updateProperties()
+        .defaultFormat(format)
+        .commit();
+  }
+
+  private EncryptedOutputFile createEncryptedOutputFile() {
+    return fileFactory.newOutputFile();
+  }
+
+  private DataFile prepareDataFile(FileAppenderFactory<Record> appenderFactory, List<Record> rowSet)
+      throws IOException {
+    DataWriter<Record> writer = appenderFactory.newDataWriter(createEncryptedOutputFile(), format, null);
+    try (DataWriter<Record> closeableWriter = writer) {
+      for (Record record : rowSet) {
+        closeableWriter.add(record);
+      }
+    }
+
+    return writer.toDataFile();
+  }
+
+  private Record createRow(Integer id, String data) {
+    Record row = gRecord.copy();
+    row.setField("id", id);
+    row.setField("data", data);
+    return row;
+  }
+
+  private StructLikeSet expectedRowSet(Iterable<Record> records) {
+    StructLikeSet set = StructLikeSet.create(table.schema().asStruct());
+    records.forEach(set::add);
+    return set;
+  }
+
+  private StructLikeSet actualRowSet(String... columns) throws IOException {
+    StructLikeSet set = StructLikeSet.create(table.schema().asStruct());
+    try (CloseableIterable<Record> reader = IcebergGenerics.read(table).select(columns).build()) {
+      reader.forEach(set::add);
+    }
+    return set;
+  }
+
+  @Test
+  public void testSortedPosDelete() throws IOException {
```

Comment:
Can you also add a test for passing a `null` row when the schema is provided to the appender factory? That should fail because when rows are passed, every row must be passed to ensure the stats are correct.

Reply:
Make sense !